### PR TITLE
fix(perf): Restrict the recent ClusterController-focused perf changes…

### PR DESCRIPTION
**Pull request by ajordens**
… to aws

Unfortunately there were some issues / inconsistencies with what an
unexpanded server group means to different cloud providers.

Any cloud provider looking to implement supportsMinimalClusters() will need
to verify the behavior of `ClusterController.getServerGroup()` and
`ClusterController.getTargetServerGroup()` as they have expectations of what
data is available in a 'minimal' cluster.

### CODE REVIEW

**Comment by ajordens** [comment]
FYI @duftler @lwander @ttomsu ... I'm just about to validate aws is ok but this is an attempt at making the previous changes Amazon-only.

**Comment by duftler** [comment]
These changes LGTM. Would also like lwander and/or ttomsu to have a look as they're more up-to-speed on the earlier changes.

File: clouddriver-web/src/test/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterControllerSpec.groovy [hunk 1]
```diff
@@ -102,29 +104,6 @@ class ClusterControllerSpec extends Specification {
       result == [serverGroup]
   }
 
-  void "should delegate to ServerGroupController when region is provided"() {
```
Comment (ajordens) [inline clouddriver-web/src/test/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterControllerSpec.groovy]: I removed this optimization entirely.

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy [hunk 2]
```diff
@@ -200,8 +192,11 @@ class ClusterController {
       throw new NotFoundException("Target not found (target: ${target})")
     }
 
+    // we can optimize loads iff the cloud provider supports loading minimal clusters (ie. w/o instances)
+    def shouldExpand = !clusterProviders.find { it.cloudProviderId == cloudProvider }.supportsMinimalClusters()
+
     // load all server groups w/o instance details (this is reasonably efficient)
-    def sortedServerGroups = getServerGroups(application, account, clusterName, cloudProvider, null /* region */, false).findAll {
+    def sortedServerGroups = getServerGroups(application, account, clusterName, cloudProvider, null /* region */, shouldExpand).findAll {
```
Comment (lwander) [inline clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy]: Maybe I'm misreading this, but for the non-aws case `supportsMinimalClusters` is false, so `shouldExpand` is true. The old default was passing in false, but now it's true. Is that expected?

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy [hunk 3]
```diff
@@ -95,4 +95,11 @@ interface ClusterProvider<T extends Cluster> {
    * @return the identifier of the backing cloud provider
    */
   String getCloudProviderId()
+
+  /**
+   * Determines whether or not optimizations can be made by retrieving minimal or unexpanded clusters.
+   *
+   * This primarily affects how server groups are loaded for a cluster (@see com.netflix.spinnaker.clouddriver.controllers.ClusterController}.
+   */
+  boolean supportsMinimalClusters()
```
Comment (anotherchrisberry) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy]: If this were a Java interface with a default method for this, would Groovy accept that and prevent you from having to touch all the cloud providers?

Not sure that's better - might be best to make it explicit.

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy (hunk 2 above)
Comment (ajordens) [reply clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy]: The pre-perf change default was true (or more specifically it wasn't defined and was true inside each cloud provider)

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy (hunk 3 above)
Comment (ajordens) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy]: That is 100% ... in the interest of not messing around I didn't convert the interface.

File: clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy (hunk 2 above)
Comment (ajordens) [reply clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy]: With the change, *only* aws should have `shouldExpand` = false.

Comment (lwander) [reply clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/ClusterController.groovy]: Ok sounds good

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy (hunk 3 above)
Comment (ttomsu) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/ClusterProvider.groovy]: Groovy has traits that are similar to java 8 default interface methods. They were handy in the cases I've used them.

**Pull request merged**